svix-cli/src/relay/mod.rs

```diff
@@ -0,0 +1,467 @@
+
+        attempt_count += 1;
+        last_attempt = Instant::now();
+
+        tokio::time::sleep(backoff).await;
+    }
+}
+
+fn receive_url(token: &str) -> String {
+    format!("https://play.svix.com/in/{token}/")
+}
+
+fn view_url(token: &str) -> String {
+    format!("https://play.svix.com/view/{token}/")
+}
+
+type S = WebSocketStream<MaybeTlsStream<TcpStream>>;
+
+struct WsConnection {
+    stream: S,
+}
+
+impl WsConnection {
+    async fn new(websocket_url: &url::Url) -> Result<Self> {
+        let request = websocket_url.to_string().into_client_request()?;
+        let (stream, _resp) = connect_async(request)
+            .await
+            .with_context(|| "failed to connect to websocket server")?;
+
+        Ok(Self { stream })
+    }
+}
+
+async fn read_from_ws_loop(
+    mut rx: SplitStream<S>,
+    tx: UnboundedSender<MessageOut>,
+    local_url: url::Url,
+    client: HttpClient,
+) -> Result<()> {
+    // We expect to see roughly _at least one Ping_ in each `SERVER_PING_PERIOD`.
+    // Other messages may arrive ahead of this schedule.
+    // Tracking the time each message is received, we can know if the server has been quiet for too
+    // long, possibly requiring us to reconnect.
+    let mut last_msg = Instant::now();
+
+    loop {
+        const REMOTE_SERVER_CLOSED: &str = "remote server closed connection";
+
+        match tokio::time::timeout(SERVER_PING_PERIOD, rx.next()).await {
+            Err(_timeout_hit) => {
+                // Generous. 1.5x the ping frequency. If we go that long without
+                // seeing anything from the server, force a reconnect.
+                if last_msg.elapsed() > SERVER_PING_PERIOD + (SERVER_PING_PERIOD / 2) {
+                    anyhow::bail!(REMOTE_SERVER_CLOSED);
+                }
+            }
+            // Stream empty/closed
+            Ok(None) => break,
+            Ok(Some(msg)) => {
+                last_msg = Instant::now();
+
+                let data = match msg? {
+                    // Control messages.
+                    Message::Close(_) => anyhow::bail!(REMOTE_SERVER_CLOSED),
+                    Message::Ping(_) | Message::Pong(_) | Message::Frame(_) => continue,
+
+                    // Messages that carry data we care to process.
+                    Message::Text(s) => s.into(),
+                    Message::Binary(bytes) => bytes,
+                };
+
+                handle_incoming_message(client.clone(), data, &local_url, tx.clone()).await;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn send_to_ws_loop(
+    mut rx: UnboundedReceiver<MessageOut>,
+    mut tx: SplitSink<S, Message>,
+) -> Result<()> {
+    while let Some(msg) = rx.recv().await {
+        tokio::time::timeout(
+            WRITE_WAIT,
+            tx.send(Message::Binary(
+                serde_json::to_vec(&msg)
+                    .expect("trivial serialization")
+                    .into(),
+            )),
```

Comment:
Why not `Message::Text` / `serde_json::to_string`?

Reply:
What's the problem with binary?

Reply:
This code can only send UTF-8 because `serde_json` only produces UTF8, that's just clearer when using those. Pretty sure that everything is going to be the same on the wire.

Reply:
Yeah, I think it's the same either way in this specific case.

It's a one-time per connection thing so there's no sense in trying to optimize (though that's not really what I was going for).